### ISSUE TITLE
Feature request: Bluetooth Discovery

### ISSUE THREAD

**Issue by cyrneko** [localsend/localsend#850]
### Is your feature request related to a problem? Please describe.

Currently it is impossible to discover devices outside of your network. This is an inconvenience and regression when compared to AirDrop, that Localsend is inspired by.

### Describe the solution you'd like

AirDrop, Nearby Share and similar implementations use Bluetooth to discover devices. It'd be useful to have in Localsend to make the experience even more seemless.

### Describe alternatives you've considered

Nearby Share / Native AirDrop exist, but lack cross platform support

### Additional context

- [ ] Depends on #189 for Wifi-Direct Connections

**Comment by MMachado05** [comment]
I'm pretty sure this is a duplicate of #144, though I'm not sure how to label it as such.

**Comment by Tienisto** [comment]
WiFi Direct is not possible on iOS I think. So we would have a solution which is just Nearby Share as an third-party app

**Comment by MMachado05** [comment]
Ahh, so if any implementation of this feature were selected, I suppose it would be Bluetooth? That functionality really is what I think would blast this project to new heights; as it stands, if I'm in a University building with poor cellular service, I can't really do anything to send files aside from just emailing it to myself, especially since I'm not on a MacOS computer to begin with.

**Comment by Tienisto** [comment]
Yes, I think Bluetooth discovery + file transmission is more reasonable. Wifi Direct is better but we are limited by OS constraints (iOS in particular).

**Comment by hellojaccc** [comment]
https://github.com/schlagmichdoch/pairdrop#paired-devices-and-public-rooms---internet-transfer
It's worth looking how pairdrop approaches complex network situations.

**Comment by Tienisto** [comment]
@hellojaccc Thanks for mentioning PairDrop. The biggest downside to PairDrop is that it's just a fork of SnapDrop: They both use WebRTC for file transmission. The problem with WebRTC is that you need a third-party server.

So if PairDrop is down, you cannot share files. You could also host your own PairDrop instance but that's not the goal of LocalSend.

**Comment by rohsaurus** [comment]
> WiFi Direct is not possible on iOS I think. So we would have a solution which is just Nearby Share as an third-party app
> 
> https://developer.apple.com/forums/thread/24661

Wouldn't it be more than Nearby Share because it would at least support Windows, Linux, and I believe MacOS? This right here is the biggest reason I can't use LocalSend right now. At least for iOS, some other solution could be figured out.

**Comment by janopae** [comment]
I think it would be reasonable to fall back to direct Bluetooth file transfer if Wifi direct is not available. The most important thing is that it doesn't fail with older devices (or iOS ones). If the user has a device with limited possibilities then they have to live with lower speed.

That said, it might be reasonable to implement this together with #144, and then later add the ability to automatically upgrade to Wifi direct when implementing #189.

So the device list would show devices found via Bluetooth, merged with devices in the same network that are too far away to be visible via Bluetooth. When starting a file transfer, it would first try to establish a connection via the common Wifi network, and if it doesn't work, fall back to Bluetooth file transfer. When after #189 being implemented, it would first try Wifi direct, then Wifi over common router, then fall back to Bluetooth.

*Enhancement idea: maybe devices only supporting Bluetooth could be shown with an indicator or in a different colour. We could use green and blue to make the distinction, just to trigger iOS users that make fun of other users not having iMessage, having the messages turn green when falling back to SMS ;)*

**Comment by MMachado05** [comment]
This implementation plan looks exciting! As likely mentioned a couple of times before, until some sort of direct device-to-device file transfer capability is implemented, LocalSend is functionally just a quicker alternative to emailing yourself a file (if we ignore privacy concerns, which the average user probably does). If I don't have a WiFi connection on either device, LocalSend doesn't work, where AirDrop does. If I'm in a building with a certain kind of public WiFi that doesn't allow device discovery, LocalSend doesn't work, where AirDrop does. If some form of Blueooth/WiFi Direct file transfer is implemented, however, then LocalSend will *really* turn into an actual AirDrop alternative, and it'll finally provide non-Apple users a reliable way to send files across devices!

**Comment by richardstefun** [comment]
Please make this happen. No other app does this, AFAIK

**Comment by evil0th** [comment]
Please make this happen. It's exactly what i need

**Comment by lwd-temp** [comment]
Find this may be helpful https://github.com/nearby-sharing/android

**Comment by taukakao** [comment]
Just chiming in to say that I plan to work on this for my bachelors thesis. 
I will probably start in October.

Of course I don't want to keep anyone from implementing this sooner.

**Comment by Eyion** [comment]
I have localsend on 2 computers with windows 10 and 1 android phone, but 2 computer pairs find each other have problems (high probability can not find each other, after constant refresh, occasionally can find each other), android phone and 2 computers can find each other.

**Comment by SMSourov** [comment]
@taukakao any update?

**Comment by taukakao** [comment]
@SMSourov 
Yes, I'm actually starting today. Though only with building it and some research. I have no ETA for when I will be done yet.

**Comment by SMSourov** [comment]
@taukakao anyway, thanks man!

**Comment by taukakao** [comment]
@MysteriousOrb2000 
I know the slow progress is frustrating but I'm still absolutely gonna work on this. (I need it for my bachelors degree after all)
Meaningful progress will be tracked in a draft PR here, so please don't ask for further updates, since this issue is for sharing ideas and concerns about the feature request.

**Comment by MMachado05** [comment]
Do we know if there are any open PR Drafts with this functionality? Just had another "my University WiFi doesn't allow device discovery" moment that made me consider this open issue.
